Update to Swift 6

fpseverino opened the pull request:
- Update package to Swift 6
- Remove all `EventLoopFuture`s in favor of async/await
- Update JWTKit to v5
- Update to Swift Testing
- Adopt `swift-format`
- Update CI

vamsii777 commented on Package.swift:
```diff
@@ -1,10 +1,10 @@
     name: "Imperial",
     platforms: [
-       .macOS(.v10_15)
+       .macOS(.v14)
```
It would be worth clarifying the reason for targeting `.macOS(.v14)` instead of the minimum required `.macOS(.v13)` by JWTKit. Is there a specific feature or improvement in `.v14` that we depend on, or could we revert to ensure broader compatibility?

fpseverino replied:
No not really, I was planning ahead for a future update to JWTKit v5 and mistakenly thought it required macOS 14

vamsii777 commented:
Just a quick note to highlight that while many parts of the codebase, such as `GoogleJWT` and `GitHub` services, leverage modern `async/await` constructs, there are still sections, like the `Imgur` service, that use legacy `Future`-based approaches.

fpseverino commented:
> there are still sections, like the `Imgur` service, that use legacy `Future`-based approaches.

Those are basically dead code not visible to the user, they're not even part of a target and are stuck in very old versions of Imperial.
Reviving them requires a bit of work, since the protocols have changed, and I think it's best to do it in a separate PR

ptoffy commented on Sources/ImperialAuth0/Auth0.swift:
```diff
@@ -2,17 +2,17 @@
     @discardableResult
     public required init(
-        routes: RoutesBuilder,
+        routes: any RoutesBuilder,
```
```suggestion
        routes: some RoutesBuilder,
```

ptoffy commented on Sources/ImperialCore/Routing/FederatedServiceRouter.swift:
```diff
@@ -43,32 +43,32 @@ public protocol FederatedServiceRouter {
     ///   - authenticateCallback: Execute custom code within the authenticate closure before redirection.
     /// - Throws: N/A
-    func configureRoutes(withAuthURL authURL: String, authenticateCallback: ((Request) throws -> (EventLoopFuture<Void>))?, on router: RoutesBuilder) throws
+    func configureRoutes(withAuthURL authURL: String, authenticateCallback: ((Request) async throws -> Void)?, on router: any RoutesBuilder) throws
```
```suggestion
    func configureRoutes(withAuthURL authURL: String, authenticateCallback: ((Request) async throws -> Void)?, on router: some RoutesBuilder) throws
```

ptoffy commented on Sources/ImperialCore/Routing/FederatedServiceRouter.swift:
```diff
@@ -77,20 +77,19 @@ extension FederatedServiceRouter {
     public var callbackHeaders: HTTPHeaders { [:] }
    
-    public func configureRoutes(withAuthURL authURL: String, authenticateCallback: ((Request) throws -> (EventLoopFuture<Void>))?, on router: RoutesBuilder) throws {
+    public func configureRoutes(withAuthURL authURL: String, authenticateCallback: ((Request) async throws -> Void)?, on router: any RoutesBuilder) throws {
```
```suggestion
    public func configureRoutes(withAuthURL authURL: String, authenticateCallback: ((Request) async throws -> Void)?, on router: some RoutesBuilder) throws {
```

ptoffy commented on Sources/ImperialCore/Services/FederatedService.swift:
```diff
@@ -42,5 +42,5 @@ public protocol FederatedService {
     ///   - completion: The completion handler that will fire at the end of the callback route. The access token is passed into the callback and the response that is returned will be returned from the callback route. This will usually be a redirect back to the app.
     /// - Throws: Any errors that occur in the implementation.
-    init(routes: RoutesBuilder, authenticate: String, authenticateCallback: ((Request) throws -> (EventLoopFuture<Void>))?, callback: String, scope: [String], completion: @escaping (Request, String) throws -> (EventLoopFuture<ResponseEncodable>)) throws
+    init(routes: any RoutesBuilder, authenticate: String, authenticateCallback: ((Request) async throws -> Void)?, callback: String, scope: [String], completion: @escaping (Request, String) async throws -> any AsyncResponseEncodable) throws
```
```suggestion
    init(routes: some RoutesBuilder, authenticate: String, authenticateCallback: ((Request) async throws -> Void)?, callback: String, scope: [String], completion: @escaping (Request, String) async throws -> any AsyncResponseEncodable) throws
```

ptoffy commented on Sources/ImperialDiscord/Discord.swift:
```diff
@@ -2,17 +2,17 @@
     @discardableResult
     public required init(
-        routes: RoutesBuilder,
+        routes: any RoutesBuilder,
```
```suggestion
        routes: some RoutesBuilder,
```

ptoffy commented on Sources/ImperialDropbox/Dropbox.swift:
```diff
@@ -2,17 +2,17 @@
     @discardableResult
     public required init(
-        routes: RoutesBuilder,
+        routes: any RoutesBuilder,
```
```suggestion
        routes: some RoutesBuilder,
```

ptoffy commented on Sources/ImperialKeycloak/Keycloak.swift:
```diff
@@ -2,17 +2,17 @@
     @discardableResult
     public required init(
-        routes: RoutesBuilder,
+        routes: any RoutesBuilder,
```
```suggestion
        routes: some RoutesBuilder,
```

ptoffy commented on Sources/ImperialCore/Helpers/Optional+Imperial.swift:
```diff
@@ -5,7 +5,7 @@ extension Optional {
     /// - Returns: The value contained in the optional.
     /// - Throws: The error passed in if the optional is `nil`.
-    public func value(or error: Error) throws -> Wrapped {
+    public func value(or error: any Error) throws -> Wrapped {
```
I'm honestly not a fan of this, I'd much rather `guard` unwrap when needed, but I'm not really precious about it. What do you think?

fpseverino replied:
I totally agree

ptoffy commented on Package.swift:
```diff
@@ -42,18 +42,27 @@ let package = Package(
+var swiftSettings: [SwiftSetting] {
+    [
+        .enableUpcomingFeature("ExistentialAny"),
+        .enableUpcomingFeature("FullTypedThrows"),
```
Are we using typed throws anywhere?

fpseverino replied:
No, not really

ptoffy replied:
We can remove it for now

ptoffy commented on Package.swift:
```diff
@@ -42,18 +42,27 @@ let package = Package(
             dependencies: [
                 .product(name: "Vapor", package: "vapor"),
                 .product(name: "JWTKit", package: "jwt-kit"),
```
I'd update to JWTKit 5, up to you whether here or in another PR

fpseverino replied:
Absolutely, it's my next priority, but I'll do it in another PR

ptoffy reviewed (changes requested):
Overall this looks good to me. There's no major changes besides a/a instead of ELFs. Besides the parameters I'd try updating the existential return types to opaque types, that should work e.g for `AsyncResponseEncodable`. I'd also add formatting

fpseverino commented:
I'm afraid that by activating formatting a lot of changes will be made that distract from the important ones. I'll do that in a separate PR

ptoffy reviewed (dismissed):
Let's also add Swift 6 language mode. I'm fine with the rest, not sure if @0xTim wants to chime in on something. I would also add some tests honestly, but that can be done later.

After thinking again I'm somewhat worried of the closures having to return opaques, as those are publicly passed in methods, but I guess we can see how that works in tests etc

fpseverino commented:
The biggest problem I have when enabling Swift 6 language mode is [this line](https://github.com/vapor-community/Imperial/blob/main/Sources/ImperialCore/Services/OAuthService.swift#L4).
If I change it to a `let` it tells me `OAuthServiceContainer` isn't `Sendable`, but it can't be `Sendable` because of the stored mutable property `services`.

Should I protect `services` with a mutex/lock or should we use something else instead of a `ThreadSpecificVariable`?

ptoffy commented:
I want to say services should be defined on the app's storage instead, which means we should be able to remove the container altogether. If that doesn't work then yeah just hide it behind a `NIOLockedValueBox`

fpseverino commented:
@ptoffy I ended up updating JWTKit to v5, adding formatting and some very very _very_ basic testing. Now it also works in Swift 6 language mode.

0xTim commented on Sources/ImperialKeycloak/KeycloakRouter.swift:
```diff
@@ -1,35 +1,36 @@
-            "redirect_uri=\(self.callbackURL)&" +
-            "scope=\(scope.joined(separator: "%20"))&" +
-            "response_type=code"
+        return "\(keycloakTokens.authURL)/auth?" + "client_id=\(self.tokens.clientID)&" + "redirect_uri=\(self.callbackURL)&"
```
Revert this

0xTim commented on Tests/ImperialTests/ImperialTests.swift:
```diff
@@ -1,10 +1,218 @@
+                    #expect(res.status == .notFound)
+                })
+
+            try app.oAuth(from: Keycloak.self, authenticate: "keycloak", callback: "keycloak-auth-complete", redirect: "/")
```
Editing routes after the app is initialised is not ideal and something we shouldn't be showcasing

0xTim commented on Tests/ImperialTests/ImperialTests.swift:
```diff
@@ -1,10 +1,218 @@
+            try await app.test(
+                .GET, "/auth0",
+                afterResponse: { res async throws in
+                    #expect(res.status != .notFound)
```
Can we do better than not 404

fpseverino replied:
Testing this is hard, but I improved it a bit, explaining with a comment why the authentication fails for each service

0xTim commented on Sources/ImperialShopify/ShopifyRouter.swift:
```diff
@@ -1,97 +1,81 @@
-    public var accessTokenURL: String = ""
+    // `accessTokenURL` used to be set inside `authURL` and read by `fetchToken`
+    // now `fetchToken` creates the `accessTokenURL` itself from the shop domain in the request
+    public let accessTokenURL: String = ""
```
Can you add a comment here that this is required by the protocol to make it clear why we have it

0xTim commented on Sources/ImperialShopify/ShopifyRouter.swift:
```diff
@@ -1,97 +1,81 @@
 
-        accessTokenURL = try accessTokenURLFrom(shop)
-        return try authURLFrom(shop, nonce: nonce).absoluteString
+        return "https://\(shop)/admin/oauth/authorize?" + "client_id=\(tokens.clientID)&" + "scope=\(scope.joined(separator: ","))&"
```
Can we split these for readability?

0xTim commented on Sources/ImperialCore/Helpers/Request+Imperial.swift:
```diff
@@ -2,35 +2,35 @@ import Foundation
         }
+
+        let token =
+            try service.tokenPrefix
+            + req
+            .accessToken()
```
These are a bit extreme!

0xTim commented on Sources/ImperialCore/Services/FederatedService.swift:
````diff
@@ -1,38 +1,36 @@
+/// Usage:
+///
+/// ```swift
+/// import HTTP
+///
+/// public class Service: FederatedService {
+///     public var tokens: FederatedServiceTokens
+///     public var router: FederatedServiceRouter
+///
+///     @discardableResult
+///     public required init(authenticate: String, callback: String, scope: [String] = [], completion: @escaping (String) -> (ResponseRepresentable)) throws {
+///         self.router = try ServiceRouter(callback: callback, completion: completion)
+///         self.tokens = self.router.tokens
+///
+///         self.router.scope = scope
+///         try self.router.configureRoutes(withAuthURL: authenticate, authenticateCallback: authenticateCallback, on: router)
+///
+///         Service.register(.service)
````
This all needs updating?

0xTim reviewed (changes requested):
Thanks! Some comments

fpseverino commented on Sources/Imperial/Services/DeviantArt/DeviantArtRouter.swift:
```diff
@@ -1,33 +1,31 @@
-            "client_id=\(self.tokens.clientID)&" +
-            "redirect_uri=\(self.callbackURL)&\(scope)" +
-            "response_type=code"
+        return "https://www.deviantart.com/oauth2/authorize?" + "client_id=\(self.tokens.clientID)&"
```
This is old code that doesn't work anymore (it's not even in a target), so a bit more work will be needed to update it.
Anyways, I'll make this change 👍🏻

ptoffy commented on Tests/ImperialTests/ImperialTests.swift:
```diff
@@ -1,10 +1,218 @@
+
+        #expect(
+            ServiceError.noServiceEndpoint("test").description
+                == "ServiceError(errorType: noServiceEndpoint, service does not have available endpoint for key: test)")
```
Can we use a more type safe approach for testing errors here?

ptoffy reviewed (commented):
This was so nicely small and reviewable a few commits ago 😅 Why are there `Future`s in some providers? E.g DeviantArt, Imgur...

vamsii777 commented on Sources/ImperialShopify/ShopifyRouter.swift:
```diff
@@ -1,97 +1,85 @@
 
-        accessTokenURL = try accessTokenURLFrom(shop)
-        return try authURLFrom(shop, nonce: nonce).absoluteString
+        return "https://\(shop)/admin/oauth/authorize?"
+            + "client_id=\(tokens.clientID)&"
+            + "scope=\(scope.joined(separator: ","))&"
+            + "redirect_uri=\(callbackURL)&"
+            + "state=\(nonce)"
```
Great work! For consistency, it's better to continue using `URLComponents` here as it's already used in other parts.

vamsii777 reviewed (commented):
Great work!

vamsii777 commented on Sources/ImperialShopify/Shopify.swift:
```diff
@@ -2,30 +2,27 @@
+    public var router: any FederatedServiceRouter { self.shopifyRouter }
 
-    public var tokens: FederatedServiceTokens { self.router.tokens }
-    public var router: FederatedServiceRouter { self.shopifyRouter }
-
-    public var shopifyRouter: ShopifyRouter
+    public let shopifyRouter: ShopifyRouter
```
Why use `var` instead of `let` since we don’t need computed properties?

vamsii777 replied:
Do we need `shopifyRouter` here?

fpseverino replied:
No we don't

vamsii777 commented on Sources/ImperialDiscord/DiscordRouter.swift:
```diff
@@ -1,22 +1,23 @@
-    public let callbackCompletion: (Request, String) throws -> (EventLoopFuture<ResponseEncodable>)
-    public var scope: [String] = []
+final public class DiscordRouter: FederatedServiceRouter {
+    public static let baseURL: String = "https://discord.com/"
```
Do we need `baseURL`?

vamsii777 commented on Sources/ImperialGitHub/GitHubRouter.swift:
```diff
@@ -1,12 +1,11 @@
-    public let callbackCompletion: (Request, String) throws -> (EventLoopFuture<ResponseEncodable>)
-    public var scope: [String] = []
+final public class GitHubRouter: FederatedServiceRouter {
+    public static let baseURL: String = "https://github.com/"
```
same here

fpseverino commented:
In the last few commits I:
- changed all `class`es to `struct`s, since I don't think any of them needed to be a reference type (I could easily be wrong)
- made a lot of `public` APIs `internal` or `package` (I think they were made before the `package` access level)
- removed `FederatedCreatable` and `OAuthService`, as their only purpose IIUC was to get data (like user and email) from the services in a very convoluted way (even the Kodeco book uses a simpler custom solution), without considering that the endpoints they provided could be outdated and require some maintenance that I think we can avoid.

Let me know what you think!

0xTim reviewed (approved):
Nothing blocking. I have thoughts about how much we're making package protected in case users want to extend Imperial to support other OAuth providers (especially internal ones that they can't upstream) but it's much easier to go that way than try and lock it down after. Lets get this into people's hands to start trying out
